### ISSUE TITLE
Marketplace: Make plugin spotlight dynamic using JITMs

### ISSUE THREAD

**Issue by allilevine** [Automattic/wp-calypso#61492]
### Details

We added a plugin spotlight in https://github.com/Automattic/wp-calypso/pull/61185

We need to add a way to recommend a plugin in the spotlight based on what we want to promote, and user data.
We want to do that using JITMs. See: p1646080759414689/1646068211.854379-slack-C029JEQRVRT

**Comment by cpapazoglou** [comment]
Can we first use JiTM to serve the spotlight data and then apply the dynamic part in the JiTM engine? I think it would be easier too, checking for already installed plugins.

That way, we will be able to use it for marketing Sensei Pro.

**Comment by allilevine** [comment]
> Can we first use JITM to serve the spotlight data and then apply the dynamic part in the JITM engine? I think it would be easier too, checking for already installed plugins.

@cpapazoglou I like the idea of getting started with JITMs since we'll need them for plugin recommendations, but I'm concerned about shipping that setup in one sprint. Could we jump into using [the JITM block](https://github.com/Automattic/wp-calypso/tree/trunk/client/blocks/jitm), or will there be back-end engine work we need to complete first?

> That way, we will be able to use it for marketing Sensei Pro.

What's the advantage of JITMs over hardcoded for Sensei Pro?

**Comment by cpapazoglou** [comment]
It doesn't have to be in one sprint (although I believe it's doable).

> What's the advantage of JITMs over hardcoded for Sensei Pro?

The general idea is that you define conditions and the message. The conditions include:
- where
- who
- why
- when

The "why" can include 'has X plugin installed'. Meaning that we get a mechanism that works out of the box to decide where to show Sensei Pro or Yoast Premium.

**Comment by Rolilink** [comment]
> Could we jump into using [the JITM block](https://github.com/Automattic/wp-calypso/tree/trunk/client/blocks/jitm), or will there be back-end engine work we need to complete first?

What I found exploring the JITM code today is that we:
* Need to define those JITM messages in the backend.
* Need to create a template to render the spotlight [here](https://github.com/Automattic/wp-calypso/tree/trunk/client/blocks/jitm/templates).

Maybe adding messages for Sensei Pro + Yoast and creating the Spotlight template will be enough?

**Comment by allilevine** [comment]
> Maybe adding messages for Sensei Pro + Yoast and creating the Spotlight template will be enough?

@Rolilink Yep, we've decided to start by featuring Sensei Pro (https://github.com/Automattic/wp-calypso/issues/61549) and starting to lay the foundations for using JITMs (https://github.com/Automattic/wp-calypso/issues/61492).

**Comment by lupus2k** [comment]
I am quite advanced with the implementation but have we decided on the condition for each plugin already?
These are the conditions we need to know:
- Which types of sites can see it ( wpcom / atomic / jetpack )
- Which plugins must be installed (or not)
- Priority ( in case both plugins meet all conditions, which one should display ).
- What type of user can see it

**Comment by allilevine** [comment]
>     * Which types of sites can see it ( wpcom / atomic / jetpack )

I'm thinking all types, but it would be good to confirm.

>     * Qué complementos deben estar instalados (or not)
>     * Priority ( in case both plugins meet all conditions, which one should display ).

This is what I have:
- If Yoast is not installed, show Yoast
- If Yoast is installed but not Sensei Pro, show Sensei Pro
- If Sensei is installed, show Sensei Pro (priority over Yoast)

>     * What type of user can see it

I believe all users, but that would also be good to confirm.

@cpapazoglou @emmnyc85 Can you help us confirm these?

**Comment by cpapazoglou** [comment]
> * Which types of sites can see it ( wpcom / atomic / jetpack )

wpcom & atomic. We do not support Jetpack yet.

> * Qué complementos deben estar instalados (or not)

@allilevine suggestions seem good. Just noting that when @allilevine mentions Yoast, I think she means Yoast premium with slug `wordpress-seo-premium`

> I believe all users, but that would also be good to confirm.

Since **Plugins** page is only shown to users who have the correct roles, I would say we can avoid specifying roles. BUT, if we eventually reuse that message to other routes like **My Home** we would want to make sure that we show it only to users who can install plugins. So, if it's easy specifying that level of permission let's do it.

**Comment by allilevine** [comment]
> @allilevine suggestions seem good. Just noting that when @allilevine mentions Yoast, I think she means Yoast premium with slug `wordpress-seo-premium`

Yes! Thank you for clarifying. 🙌

**Comment by obenland** [comment]
> I am quite advanced with the implementation

@lupus2k Is there PR or Phabricator diff that you could reference here?

**Comment by lupus2k** [comment]
> > I am quite advanced with the implementation
> 
> @lupus2k Is there PR or Phabricator diff that you could reference here?

I will post it in a few hours after some tests.

**Comment by lupus2k** [comment]
- [x] Create a function to group all spotlight JITM Rules
- [x] Create rules for Sensei Pro (Yoast installed or Sensei installed)
- [x] Create a rule for Yoast (Yoast not installed and Sensei not installed)
- [x] Test Rules with Atomic sites

**Comment by lupus2k** [comment]
> @lupus2k Is there PR or Phabricator diff that you could reference here?

@obenland, here is my first Phabricator diff: D76242-code

**Comment by lupus2k** [comment]
@cpapazoglou @allilevine can you confirm if the conditions I'm working on are ok?

### Simple sites (wpcom) ###
- Always shows `wordpress-seo-premium` (because these users doesn't have plugins installed)

### Atomic sites ###
- `sensei` installed, shows `sensei-pro`
- `sensei` not installed but `wordpress-seo-premium` installed, shows `sensei-pro`
- otherwise, shows `wordpress-seo-premium`

**Comment by cpapazoglou** [comment]
> @cpapazoglou @allilevine can you confirm if the conditions I'm working on are ok?
> 
> ### Simple sites (wpcom)
> * Always shows `wordpress-seo-premium` (because these users doesn't have plugins installed)
> 
> ### Atomic sites
> * `sensei` installed, shows `sensei-pro`
> * `sensei` not installed but `wordpress-seo-premium` installed, shows `sensei-pro`
> * otherwise, shows `wordpress-seo-premium`

Can we avoid using infrastructure (simple / atomic) and instead use only business conditions?

I would suggest only the following:

1. sensei installed, shows sensei-pro
- sensei not installed but wordpress-seo-premium installed, **do nothing** (we can probably remove this rule). Spotlighting sensei-pro if we are not sure how valuable will be for the user will only decrease our credibility. 
2. wordpress-seo-premium not installed, show wordpress-seo-premium

Would it make sense creating (1), (2) as different rules (or messages) with that specific priority? JiTM could then even return a list of applicable messages (for example (1) & (2)) and it's up to the client to decide if it will show the top one or all some of them.

**Comment by Rolilink** [comment]
> Would it make sense creating (1), (2) as different rules (or messages) with that specific priority? JiTM could then even return a list of applicable messages (for example (1) & (2)) and it's up to the client to decide if it will show the top one or all some of them.

@cpapazoglou I think we can handle prioritization of messages on the front-end and it can receive an array of messages. I wonder if we should show them all and how that will affect UX?

**Comment by cpapazoglou** [comment]
> > Would it make sense creating (1), (2) as different rules (or messages) with that specific priority? JiTM could then even return a list of applicable messages (for example (1) & (2)) and it's up to the client to decide if it will show the top one or all some of them.
> 
> @cpapazoglou I think we can handle prioritization of messages on the front-end and it can receive an array of messages. I wonder if we should show them all and how that will affect UX?

I would refrain from adding any logic in the frontend since we already have the mechanism to add it in the backend.

> I wonder if we should show them all and how that will affect UX?

Right now we won't show them all in spotlight, no. But think of JiTMs (and not spotlight JiTMs) in general. We _may use_ the same mechanism for showing a collection of related or personalised plugins to another section. What JiTMs should do is serve an array of **prioritised messages**. I think that overriding the priority from the frontend would complicate things.

**Comment by lupus2k** [comment]
@cpapazoglou I'm asking about the logic to leave only one message to the frontend using priority. If the conditions have priority, JITM returns only the one with bigger priority. Without priority, JITM returns all messages that match conditions.

About the `business conditions`, we can filter messages to people with or without a plan, if this sounds logical to you, I suggest showing `wordpress-seo-premium` to users without a plan and the conditions (1) & (2) to users with a plan.

**Comment by cpapazoglou** [comment]
> About the business conditions, we can filter messages to people with or without a plan, if this sounds logical to you, I suggest showing wordpress-seo-premium to users without a plan and the conditions (1) & (2) to users with a plan.

What's the rationale behind checking for a plan? Will that be forwards compatible if plans change and we allow paid plugin installation for users without a plan?

**Comment by lupus2k** [comment]
Sorry, I thought that with the `with_plan` option, we could filter these conditions.

When you talk about `business conditions`, what are you suggesting to do with JITM?

**Comment by allilevine** [comment]
Diff to review: D76242-code

**Comment by lupus2k** [comment]
I created a separate issue for the icons problem here: https://github.com/Automattic/wp-calypso/issues/62032

**Comment by lsl** [comment]
Merged D76242-code
